src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1201 @@
+                                           Ykt_diff.col(0), Skt_diff.col(0));
+  } else {
+    alpha_mat.col(0).setOnes();
+  }
+  for (Eigen::Index iter = 1; iter < diff_size; iter++) {
+    if (check_curve_vec[iter]) {
+      alpha_mat.col(iter) = internal::form_diag(
+          alpha_mat.col(iter - 1), Ykt_diff.col(iter), Skt_diff.col(iter));
+    } else {
+      alpha_mat.col(iter) = alpha_mat.col(iter - 1);
+    }
+  }
+  std::mutex update_best_mutex;
+  internal::debug::post_lbfgs(logger, update_best_mutex, param_size,
+                              num_elbo_draws, alpha_mat, Ykt_diff, Skt_diff);
+  auto lp_fun = [&model](auto&& u, auto&& streamer) {
+    return model.template log_prob<false, true>(u, &streamer);
+  };
+  // NOTE: We always push the first one no matter what
+  check_curve_vec[0] = true;
+  std::vector<boost::ecuyer1988> rng_vec;
+  const int tbb_threads = tbb::this_task_arena::max_concurrency();
+  rng_vec.reserve(tbb_threads);
+  for (Eigen::Index i = 0; i < tbb_threads + 1; i++) {
+    rng_vec.emplace_back(
+        util::create_rng<boost::ecuyer1988>(random_seed, path + i));
+  }
+  Eigen::Index best_E = -1;
+  internal::elbo_est_t elbo_best;
+  internal::taylor_approx_t taylor_approx_best;
+  std::atomic<size_t> num_evals{lbfgs.grad_evals()};
+  tbb::parallel_for(
+      tbb::blocked_range<Eigen::Index>(0, diff_size),
+      [&](tbb::blocked_range<Eigen::Index> r) {
+        auto&& thread_rng
+            = rng_vec[tbb::this_task_arena::current_thread_index()];
+        for (int iter = r.begin(); iter < r.end(); ++iter) {
+          std::string iter_msg(path_num + "Iter: [" + std::to_string(iter)
+                               + "] ");
+          if (STAN_DEBUG_PATH_ITERS) {
+            logger.info(iter_msg + "\n------------ Iter: "
+                        + std::to_string(iter) + "------------\n");
+          }
+          auto alpha = alpha_mat.col(iter);
+          std::vector<Eigen::Index> ys_cols;
+          ys_cols.reserve(history_size);
+          {
+            for (Eigen::Index end_iter = iter; end_iter >= 0; --end_iter) {
+              if (check_curve_vec[end_iter]) {
+                ys_cols.push_back(end_iter);
+              }
+              if (ys_cols.size() == history_size) {
+                break;
+              }
+            }
+          }
+          const auto current_history_size = ys_cols.size();
+          std::vector<decltype(Ykt_diff.col(0))> Ykt_h;
+          Ykt_h.reserve(current_history_size);
+          Eigen::MatrixXd Skt_mat(Skt_diff.rows(), current_history_size);
+          for (Eigen::Index i = 0; i < current_history_size; ++i) {
+            Ykt_h.push_back(Ykt_diff.col(ys_cols[i]));
+            Skt_mat.col(i) = Skt_diff.col(ys_cols[i]);
+          }
+          Eigen::VectorXd Dk(current_history_size);
+          for (Eigen::Index i = 0; i < current_history_size; i++) {
+            Dk.coeffRef(i) = Ykt_h[i].dot(Skt_mat.col(i));
```

Comment:
Yes good catch! I'll remove this computation, but I'm going to leave `Dk` as `Eigen::VectorXd Dk = Rk.diagonal();` since we iterate over `Dk` a good bit and it would be nice for those values to all be contiguous